File: benchmarks/clojure/mandelbrot.clj
(ns mandelbrot-benchmark
  (:gen-class))

;; Size constants
(def width 800)
(def height 800)
(def max-iterations 100)

;; Calculate the Mandelbrot set
(defn calculate-mandelbrot [max-iterations]
  (vec (for [y (range height)]
    (vec (for [x (range width)]
      (let [cx (/ (* (- (double x) (/ width 2.0)) 4.0) width)
            cy (/ (* (- (double y) (/ height 2.0)) 4.0) height)]
        (loop [zx 0.0
               zy 0.0
               iteration 0]
          (if (and (<= (+ (* zx zx) (* zy zy)) 4.0) (< iteration max-iterations))
            (let [temp (+ (- (* zx zx) (* zy zy)) cx)]
              (recur temp (+ (* 2.0 zx zy) cy) (inc iteration)))
            iteration)))))))))

;; Count non-black pixels in the result
(defn count-non-black [result max-iterations]
  (reduce + (for [y (range height)]
               (count (filter #(< % max-iterations) (nth result y))))))

(defn -main [& args]
  (let [start-time (System/currentTimeMillis)]
    
    (let [result (calculate-mandelbrot max-iterations)
          count (count-non-black result max-iterations)]
      
      (println "Mandelbrot set calculation finished.")
      (println (str "Image size: " width " x " height))
      (println (str "Maximum iterations: " max-iterations))
      (println (str "Non-black pixels: " count))
      
      (println (str "Time taken: " (- (System/currentTimeMillis) start-time) " ms"))
      
      ;; Estimate memory usage
      (println (str "Memory used: " (/ (- (Runtime/getRuntime) (.freeMemory (Runtime/getRuntime))) 1024) " KB")))))

File: benchmarks/haskell/mandelbrot.hs
-- Mandelbrot set calculation benchmark for Haskell

import System.Environment
import System.IO
import Control.Monad
import Data.Time.Clock
import Text.Printf
import Data.Complex
import qualified Data.Vector as V
import qualified Data.Vector.Unboxed as UV

-- Constants
width, height :: Int
width = 800
height = 800

maxIterations :: Int
maxIterations = 100

-- Check if a point belongs to the Mandelbrot set
isInMandelbrot :: Complex Double -> Int -> Int
isInMandelbrot c maxIter = go 0 (0 :+ 0)
  where
    go i z
      | i >= maxIter = maxIter
      | magnitude z > 2 = i
      | otherwise = go (i + 1) (z*z + c)

-- Calculate the entire Mandelbrot set
computeMandelbrot :: Int -> V.Vector (UV.Vector Int)
computeMandelbrot maxIter = V.generate height $ \y ->
  UV.generate width $ \x -> isInMandelbrot (toComplex x y) maxIter
  where
    toComplex x y = (fromIntegral x - fromIntegral width / 2) * 4 / fromIntegral width :+
                    (fromIntegral y - fromIntegral height / 2) * 4 / fromIntegral height

-- Count non-black pixels
countNonBlack :: V.Vector (UV.Vector Int) -> Int -> Int
countNonBlack result maxIter = V.sum $ V.map (UV.length . UV.filter (< maxIter)) result

-- Main function
main :: IO ()
main = do
    startTime <- getCurrentTime
    
    -- Compute the Mandelbrot set
    let result = computeMandelbrot maxIterations
    let nonBlackCount = countNonBlack result maxIterations
    
    putStrLn $ "Mandelbrot set calculation finished."
    printf "Image size: %d x %d\n" width height
    printf "Maximum iterations: %d\n" maxIterations
    printf "Non-black pixels: %d\n" nonBlackCount
    
    endTime <- getCurrentTime
    let diff = diffUTCTime endTime startTime
    printf "Time taken: %s ms\n" (show (diff * 1000))